### ISSUE TITLE
Plugin Mechanism

### ISSUE THREAD

**Issue by delvh** [go-gitea/gitea#20126]
## What should it be?

- it should allow adding additional functionality to Gitea without bloating the main binary or being responsible for maintaining it
- it should **not** allow to modify the core functionality of Gitea

## Architecture

- ~~Plugins are called using `Go`s `plugin` package.~~ (see discussion below, was generally disliked)
- Plugins are called over _RCP_ using the [go-plugin](https://github.com/hashicorp/go-plugin) package.
- Reasons:
  - active community to maintain the package
  - works on every platform and even across languages
  - minimalistic and understandable -> no technical debt to have to consider
  - OS is responsible for setting plugin resource limits, not Gitea
  - Protocol versioning built-in
- Reasons against that architecture:
  - requests over RCP make everything slow
  - extra dependency
  - vendor lock-in for plugins to `go-plugin`

### Communication: _Gitea :arrow_right: plugin :arrow_right: Gitea_

- common data structures and methods needed
- recommendation: an extra package that both Gitea and the plugins use
  - during application startup, Gitea should use the package from the receiver side and the plugins from the sender side
  - while dispatching calls to the plugins, the roles should be switched around so that Gitea becomes the sender and the plugins the receivers
- things that plugins, as well as Gitea, should have access to (at best only as a wrapper that is set once at runtime because the underlying implementation can change quite often):
  - especially things located in `modules/`:
  - logger
  - context
  - i18n
- again, to avoid vendor lock-in, this package **cannot** depend on anything that it doesn't define on its own, or is provided by Go itself
- to reduce the chance of plugins breaking with a new Gitea release, this extra package should be developed to be "all-inclusive" from the beginning meaning that it entails everything a plugin might wish to configure
  - it is recommended that the extra package uses easily extendable data structures, i.e. `map[string]?`, where possible to avoid breaking changes

### Communication: _plugin :arrow_right: Gitea :arrow_right: plugin_
- Let Plugins use the Gitea API when requesting additional information from Gitea. Reasons:
  - easy to implement (literally **NO** work at all :smiley:  )
  - easy to maintain (again, because there is literally **NO** code to maintain :smiley:  )
  - safe to use for the plugins because the API guarantees stability
  - API gets used more, hence more people want to see it well-maintained and there is an actual incentive to add an API for a feature that at that moment only exists for the web UI
- Reasons against this approach:
  - slow, because it will always require HTTP requests
  - plugins need to parse the answer instead of simply being able to use a predefined method/ datatype

### Plugin Lifecycle (What the plugins can set/ "use")
- [ ] Plugin Initialization
- [ ] `Application is completely initialized and running` - Hook
- [ ] Plugin Dependencies (to Gitea (versions), to other plugins?)
- [ ] Setting Plugin-wide settings as well while parsing the server-wide settings (requires (a) section(s) reserved for plugins)
- [ ] Custom translation keys to allow for localized new features
- [ ] Adding custom tabs to different units:
  - most basic:
  - user
  - org
  - user+org
  - all repos
  - public repo
  - private repo
  - additionally possible:
  - issue
  - PR
  - projects
  - packages
  - ...
  - to make parsing this easier, it would probably be good to use an approach that allows setting a list of tags where this tab will be shown. Problem: How to handle `OR`/ `AND` relations (depending on the chosen algorithm)?
- [ ] Adding consistency checks
- [ ] Cron Job Hooks?
- [ ] Shutdown Hook

### What data is needed by every/ some plugin(s)?
- [ ] unique plugin name (i.e. by following reverse naming such as `io.gitea.example_plugin`)
- [ ] human printable plugin name (i.e. `Example Plugin V1`)
- [ ] plugin version in semver form
- [ ] list of core units affected by this plugin (rather as a code of honor the plugin intends to keep than a security feature)
- [ ] dependencies to other plugins
- [ ] custom translations/ translation files
- [ ] custom settings/ setting files
- [ ] custom web routes
- [ ] custom API routes (problem: how to update the swagger file with the new operations?)
- [ ] provided hooks
- [ ] minimum Gitea version the plugin needs

## Why this long text rambling?
1. Everyone should know without a doubt how it should look like in the end 
1. To make the work for the (eventual) reviewers easier by simply letting them cross-check the actual functionality with its intention
1. Most of the text I wrote here can be copy-pasted into the docs, making it easier for me to write the docs
1. To allow for discussion if I missed something while planning this feature
1. To see if the scope others expect from this feature differs from what I imagine from it

**Comment by 42wim** [comment]
I think https://github.com/hashicorp/go-plugin should also be taking in consideration, afaik the "normal" plugin system has a lot of shortcomings (eg crashing the whole process because of issues with the plugin), that the Hashicorp one does not suffer from.  

See also:
- https://www.youtube.com/watch?v=SRvm3zQQc1Q
- https://alperkose.medium.com/things-to-avoid-while-using-golang-plugins-f34c0a636e8

**Comment by delvh** [comment]
Yeah, I've also seen their implementation when looking for a plugin system.
However, to me, it appears as if they don't support executing dynamic functions as I think I read the sentence `sending functions is not possible` anywhere, and then I stopped reading and didn't investigate further as that is something we most likely need.
After having read through their docs for a little, it seems I was a bit too fast with that.

Regarding the unexpected crashes: I intended to call every plugin by using something like 
```go
func PluginCall(function func(plugin *Plugin)err) {
  for _, plugin := range plugins {
    executePluginCall(plugin)
}

func executePluginCall(plugin *Plugin, function func(plugin *Plugin)err) {
  defer func(){
    if err := recover(); err != nil { // Error handling
    }
  }()
  function(plugin)
}
```
and thus mitigating the potential for those errors.

So, the question is now, should we use the RPC version which might offer a few benefits and higher latency, or do we stay with the Go native approach?

**Comment by Gusted** [comment]
> However, to me, it appears as if they don't support executing dynamic functions as I think I read the sentence `sending functions is not possible` anywhere, and then I stopped reading and didn't investigate further as that is something we most likely need.

Do you want to compile/execute Go-code on-the-fly?

**Comment by delvh** [comment]
Not really, but we certainly need to handle the custom routes which can best be achieved dynamically, and making this interface-based might be possible, but (a little) bit more work.

**Comment by Gusted** [comment]
Wish you many luck with this advanced logic.

**Comment by lunny** [comment]
I prefer the plugin system based on javascript or nodejs because that will allow it include frontend code.

Possbile library be used. https://github.com/dop251/goja, there is also a plugin support nodejs .

**Comment by delvh** [comment]
Interesting how many different views there are regarding this feature.
Now we have three completely different options and no agreement on what to use...

**Comment by lunny** [comment]
> Interesting how many different views there are regarding this feature. Now we have three completely different options and no agreement on what to use...

Use javascript based plugin system, we can implement a plugin market and one click installation. That's impossible for Go based plugin system.

**Comment by delvh** [comment]
The problem I can see is: We do not even have the same definition of what we want in the end:
What you want seems to be a per-user plugin mechanism, what I thought of was a per-instance plugin mechanism...

**Comment by lunny** [comment]
> The problem I can see is: We do not even have the same definition of what we want in the end: What you want seems to be a per-user plugin mechanism, what I thought of was a per-instance plugin mechanism...

I don't think so. I think I'm also talking about per-instance plugin. If I made you confusing, I'm sorry.

**Comment by jolheiser** [comment]
-1 for the go `plugin` package. 
It can't be removed per compatibility promise, but it is absolutely not stable, and there are far too many caveats to get it working properly.

**Comment by lunny** [comment]
I think we can begin from a theme-kind plugin, which is simple and useful. A plugin could be a zip/tar file with a `main.js` or `index.js` and other files like `theme.sass`, `theme.css` and `images` and etc.

**Comment by wxiaoguang** [comment]
I do not think the Go's `plugin` package is the future. Is there any success story of using it?

Most large Go projects uses gPRC for inter-process commutation, including plugins, but it's also heavy.

**Comment by delvh** [comment]
Ok, seems as if the `plugin` package is generally disliked, so I'll exclude it from the possible options.

I'm still not a fan of a purely frontend-based plugin mechanism as that will have **many** shortcomings of its own, and I don't know if 
> plugin market and one click installation.

should be the deciding factor for how to implement the plugin system...
Server admins should be sure of what they want for their instance, and not install it with a simple misplaced click.
Also, even with the Go-based approach, a plugin "market" would be possible as a third-party tool that simply stores the executables and perhaps some other information about the plugin...

So, as I can see that now, the two possible options are
- a backend-based plugin mechanism (most likely using `go-plugin`)
- a frontend-based plugin mechanism (no idea how to implement that -  I didn't see anything practical in the link you sent, @lunny ?)
- Did I miss something else?

**Comment by luwol03** [comment]
Isn't this library lunny suggested meant for a backend-based plugin mechanism? But instead of writing go you can write js? Or would that be to difficult to build a bridge between the Gitea go code and js code?

**Comment by lunny** [comment]
Yes, it could expose structs, methods to frontend. i.e. If we expose getdb method, then we can operate database with javascript.

**Comment by lafriks** [comment]
I think the best way to go would be using grpc so that plugins could be implemented using any language. Plugin could return asset files on init (JavaScript etc) that would need to be injected/used in specific places

**Comment by Mai-Lapyst** [comment]
Honestly, I would like to see the native go plugin implementation.
Yes it has drawbacks, but it would be fast and admins had full controll over what to use.

But since thats off the table for now, I would rather go with the gRPC approach, since its more open (i.e. you can choose more languages) and imo more secure then javascript.

The problems I see with js plugins is, that it lead to a huge security risk, since js can be so minified or encrypted that you not even need any letters (only parentheses) to write correct js. We then would need an very strict system so no js can accidentally just execute "DROP ALL TABLES;" (or similar) on startup.

Also an interpreter (of any kind) in gitea would bring back the risk of one plugin crashing the whole instance, or to eat up so much performance that gitea stops responding (i.e. a infinite loop in a script somewhere). To mitigate this we then would need to sandbox the plugins into own processes, which would nearly the same as the gRPC approach but with less freedom.

**Comment by delvh** [comment]
To be fair, you can also `DROP ALL TABLES;` in Go or consume all available memory.
But yeah, I can see your point.

**Comment by luwol03** [comment]
Then you have to think twice about an official plug-in market like lunny suggested. Official also indicates that they are save for usage and have no malicious intentions. An easy one click install could potentially break servers from not experienced users who don't understand what the plugin is doing and just pressing install. 

To prevent this, this would at least requires some reviewing and analysis of newly published plugins, which is time expensive.

**Comment by lafriks** [comment]
We could probably get some ideas on how mattermost have implemented their plugin system.

**Comment by jolheiser** [comment]
> We could probably get some ideas on how mattermost have implemented their plugin system.

For posterity:
[mattermost plugin package](https://github.com/mattermost/mattermost-server/tree/master/plugin)
It looks like they use [go-plugin](https://github.com/mattermost/mattermost-server/blob/5544e753e89d990ee87d63eae8082beb65ba0778/go.mod#L31)

**Comment by Trendyne** [comment]
Since nobody has mentioned it yet, projects like Adobe Lightroom, MPV, or Wireshark use Lua for this. You'd read and run Lua files from a directory where they would register events with callbacks or maybe return properties and functions. It's very light. simple, and fast but would make it very easy to make plugins and shouldn't be any harder to implement

Made up possible example:
```lua
-- Greet users on their first repo
gitea.on("newRepo", function(repo)
	if #repo.owner.repos == 1 then
		local msg = ("Welcome to our Gitea, %s!"):format(repo.owner.name)
		repo.owner:sendNotifacation(msg)
	end
end)


```

**Comment by KN4CK3R** [comment]
The Lua integration could be plugin too which forwards callbacks to Lua.

**Comment by l-georg** [comment]
Is there any updates on that front?

**Comment by Lord-Leonard** [comment]
Are there any updates on this

**Comment by lunny** [comment]
For most backend tasks, Gitea Actions can be considered as somewhat plugins.

**Comment by lunny** [comment]
Looks like Typescript 7 will be migrated to Go. So that maybe we can use typescript as the plugin language.

**Comment by wxiaoguang** [comment]
> Looks like Typescript 7 will be migrated to Go. So that maybe we can use typescript as the plugin language. https://github.com/microsoft/typescript-go

`typescript-go` is just a TypeScript compiler implemented in Go with a LSP. Where do you see that "use typescript as the plugin"?

----

To be frank:

For the "plugin" proposal, I do not see anyone is really working on it, or whether it is feasible at current stage (my answer is "no").

So I would suggest to mark the proposal as "close as not planned" since it doesn't seem to get real progress.

**Comment by wxiaoguang** [comment]
Quote above: So I would suggest to mark the proposal as "close as not planned" since it doesn't seem able to get real progress (reopen at any time when there is manpower)

Golang eco-system is naturely not easy with plugins ....  unfortunately.

**Comment by hiifong** [comment]
Maybe we can learn from git's plugin model, where commands that start with git- can be called freely by git, e.g. I have a git-bug plugin, and I can call my git-bug plugin by calling git bug.

**Comment by Sharaf5** [comment]
most top named Go/C/C++ projects support plugins including
- CoreDNS : pre-compiled, selectively activated Go plugins
- Traefik : Transpiling GO on the fly ( not recommended for performance ), and WASM ( may be considerd )
- Nginx : pre-shared lib `.so`
- PgSQL : plugable extensions ( citus, timescaleDB, ... )
- WebHooks could be considerd a plugin mechanism for both front and backend
- Blender, GIMP, ... had plugin systems

there were many success stories and a cost effective announce using plugins opening the way to be integrated into other projects
Really, closing it disappointed me, but hope this is in mind for future at least
I wished this project could grow beyond it's limit
hope you best luck

**Comment by wxiaoguang** [comment]
> most top named Go/C/C++ projects support plugins including

These examples you shown, are either desktop/traffic-server or C++ applications (Golang plugin doesn't work on Windows, but Gitea still needs to support), or the plugin doesn't need to heavily depends on the host application (traffic server: CoreDNS, Traefik). And either they have a large company & team, or a lot of active contributors to maintain.

For Gitea's plugin: Gitea itself is a heavy web application (but not a traffic server), there are more business logic than others, the proposed "Plugin Lifecycle"  is already more complex than the traffic server examples. And Gitea community is not that large as others (that's what I meant "my answer is no: either gPRC or js/wasm/lua would involves a lot of work to design the structs and interfaces, no manpower").

----

> WebHooks could be considerd a plugin mechanism for both front and backend

Actually it doesn't need a general plugin system, but you see that the webhook related PRs are still stale for long time (WIP: Custom Webhooks #19307). As simple as this one doesn't get success, why a general plugin proposal would succeed?

I don't want to disappoint anyone, but I think we need to face reality.

And see my comment above: reopen at any time when there is a feasible design and manpower

**Comment by wxiaoguang** [comment]
Let's keep it open and don't disappoint anyone. Let's hope there would be some real progresses.

**Comment by hiifong** [comment]
> Maybe we can learn from git's plugin model, where commands that start with git- can be called freely by git, e.g. I have a git-bug plugin, and I can call my git-bug plugin by calling git bug.

A successful case of a go project: [github cli](https://github.com/cli/cli)

**Comment by wxiaoguang** [comment]
> > Maybe we can learn from git's plugin model, where commands that start with git- can be called freely by git, e.g. I have a git-bug plugin, and I can call my git-bug plugin by calling git bug.
> 
> A successful case of a go project: [github cli](https://github.com/cli/cli), But this plugin mode only supports command line calls

You need to find some "rich web application" to learn.

**Comment by ProIcons** [comment]
There are several implementations / architectures of "Plugins". 
The most common one (not necessarily in golang) is dynamic library loading, in which, a plugin is compiled against an ABI, and it gets loaded dynamically on runtime from a predefined folder/structure. For that architecture to even exist, it requires a good isolation layer and abstraction layer. Plugins while being able to interact with the main Application, should do so from the exposed APIs to the Plugin, based also on the permissions/scope each plugin can get. It shouldn't be able to modify the database of the core application directly, or any of the behavior of the application directly without accessing this shared controlled layer. This obviously creates complexity. However, with this approach a plugin can do more on the Application, such as declaring its own http api endpoints on the same server, injecting new UI Elements, having its own database entities/state and have a tighter integration with the platform in general.

The one suggested above (as in the first message of the post), resembles more an "Integration" layer, rather than a "Plugin" layer. In this approach there are limitations, which, some of them, can be addressed with "creative workarounds" while some others can't. One of them would probably be, injecting new functionality in the UI, and having API Endpoints to handle the new functionality on the backend. (It could be solved by a third party server, but this would lead to a CORS mess, and this doesn't feel like a plugin system anymore)

An example for a plugin would be "Gitea Pages". What would it require? 
- Repo Settings (UI/API/DB) (like Github is doing, to probably select the branch from which you want to deploy pages from)
- Org Settings (UI/API/DB) (from which repo you want to deploy pages from)
- Access to artifacts (to push compilation results of static assets to the pages storage/server/cache)
- Administration Settings (UI/API/DB) (to configure Pages related settings)
- (Probably many more things, but this isn't a post about Gitea Pages rather than the Plugin architecture..)

While Gitea could serve pages, it doesn't have to. Since we're talking about a plugin after all, the whole project, besides the plugin, can have other executable services, which the plugin can interact with, in the same way Gitea interacts with Action Runners. Some executable services such as `Pages Servers` which could serve content from a Distributed Storage such as S3, and the plugin to directly store these artifacts (hugo/jekyll compilation units) directly to S3. 

For this example, which involves many systems and injection points, the dynamic library loading seems a more appropriate solution. 

Having said all that, I'm aware of the Complexity of this idea/architecture, since I have already engineered something similar in an Enterprise Product, and it is not something that can be hacked together in a week. 


> > > Maybe we can learn from git's plugin model, where commands that start with git- can be called freely by git, e.g. I have a git-bug plugin, and I can call my git-bug plugin by calling git bug.
> > 
> > 
> > A successful case of a go project: [github cli](https://github.com/cli/cli), But this plugin mode only supports command line calls
> 
> You need to find some "rich web application" to learn.

There are a lot, which are written in different languages, which handle dynamic library loading significantly different. I have exposure to mechanisms like these mostly for Languages such as Java and C#.

#### For C# 
- .NET Framework (The Old Legacy one) you have [AppDomains](https://learn.microsoft.com/en-us/dotnet/framework/app-domains/application-domains). 
- .NET (The Cross platform one) you have [AssemblyLoadContexts](https://learn.microsoft.com/en-us/dotnet/core/dependency-loading/understanding-assemblyloadcontext)

These are both Library isolation layers, which can be loaded and unloaded on runtime. There are several .NET Applications that are utilizing them such as 
- nopCommerece [Docs](https://docs.nopcommerce.com/en/developer/tutorials/description-of-plugin-system.html) [Plugin Related Tree](https://github.com/nopSolutions/nopCommerce/tree/23eacb1f981b050e8d1fa7e5922e5b2bb515d4c9/src/Libraries/Nop.Services/Plugins)
- Jellyfin [Docs](https://jellyfin.org/docs/general/server/plugins/) [Plugin Related Tree](https://github.com/jellyfin/jellyfin/tree/6d4efe6523d925398d64a889321f8cdc21eea915/MediaBrowser.Common/Plugins)
- Umbraco CMS [Docs](https://docs.umbraco.com/umbraco-cms/extending/packages/creating-a-package)
- ABP Framework Plug-in System [Docs](https://abp.io/docs/latest/framework/architecture/modularity/plugin-modules)
- BepInEx Modding Framework for Unity [Repository](https://github.com/BepInEx/BepInEx)


#### For Java
- Java 8+ you have [ClassLoaders](https://docs.oracle.com/javase/8/docs/api/java/lang/ClassLoader.html) and [ServiceLoaders](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html) 

You can unload a ClassLoader on runtime, which a ServiceLoader is using. There are also libraries that are utilizing these mechanisms such as
- PF4J [Web Page](https://pf4j.org/)

Jetbrains' IDEs are all using a Plugin System you can find all their plugins in their repository https://github.com/JetBrains/intellij-plugins

In java there are many closed-source applications that are utilizing a plugin system, which there aren't source examples for them, however they tend to use directly the Language's APIs instead of relying on a 3rd party library such as the aforementioned PF4j library.

#### For C++ 
The only exposure I have with dynamic library loading, is modding frameworks for HLDS and SRCDS Servers. The frameworks are [AmxmodX](https://github.com/alliedmodders/amxmodx) and [SourceMod](https://github.com/alliedmodders/sourcemod). HLDS and SRCDS (Valve's Dedicated Servers for GoldSrc (CS1.6, HL,...) and Source games (CS:S, CS:GO, TF2,...) respectively) are dynamically loading Amxmodx and Sourcemod on startup, and these provide an "Easier" to use abstraction layer of the intrinsics of the Dedicated Servers by loading other plugins either in their own pre-compiled intermediate language, or directly with C++. There are projects on top of them which allows Python and C# to be used for plugins.

Edit: 
While researching I found the following go package: https://pkg.go.dev/plugin
